### PR TITLE
Testbench server wip

### PR DESCRIPTION
**Pull request by touilleMan**
# What has changed ?

Start the test server:
```shell
python tests/scripts/run_testbed_server.py --port 8080 --orga-life-limit 1
```

Then start a test orga (the server create this orga by cloning an existing one so this is super fast ^^):
```shell
curl -X POST http://localhost:8080/test/new -H 'Content-Type: application/json' -d '{"kind": "coolorg"}' -v
```

After 1s the test environment gets destroyed.

### CODE REVIEW

File: oxidation/libparsec/crates/testbench/src/coolorg.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        SigningKey::from(hex!(
+            "571d726cc5586b4bfd5b20d8af2365cf8bb8c881b4925794e6e38cdcc5ec82ef"
+        )),
+        SecretKey::from(hex!(
+            "fd64fba009dc635af6662f45753b8c9772ff6e214e203e9d61ce029550e250f7"
+        )),
+        CertificateSignerOwned::User(alice1_id.clone()),
+        &alice1_signkey,
+        dt,
+    ));
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/testbench/src/coolorg.rs:97]: We could create a macro for that :thinking:

File: oxidation/libparsec/crates/client_connection/Cargo.toml [hunk 2]
```diff
@@ -4,6 +4,11 @@ version = "0.1.0"
 edition = "2021"
 autotests = false
 
+[features]
+test-testbed-support = [
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/client_connection/Cargo.toml:8]:
```suggestion
testbed-support = [
```

the `test` prefix is redundant for me

File: oxidation/libparsec/crates/platform_device_loader/src/testbed.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                .or_insert(Box::new(PseudoPersistentStorage::new(
+                    config_dir,
+                    &env.organization_addr,
+                    env.template.as_ref(),
+                )));
+        let store = store
+            .downcast_mut::<PseudoPersistentStorage>()
+            .expect("Unexpected pseudo persistent storage type for platform_device_loader");
+        cb(store)
+    })
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/platform_device_loader/src/testbed.rs:84]: Is the callback really needed ?
Why not directly return an `Option<PseudoPersistentStorage>` and use `map` on other functions ?

File: oxidation/libparsec/crates/testbed/src/env.rs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+pub enum TestbedKind {
+    ClientServer,
+    ClientOnly, // Server is considered offline for the whole run
+}
+
+#[derive(Debug)]
+pub struct TestbedEnv {
+    pub kind: TestbedKind,
+    // Config dir is used as the key identifying the testbed env
+    // note testbed is designed to always run mocked in memory
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/testbed/src/env.rs:18]:
```suggestion
    // Note: testbed is designed to always run mocked in memory.
```

File: oxidation/libparsec/crates/testbed/src/env.rs [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+#[derive(Debug, PartialEq)]
+pub enum TestbedKind {
+    ClientServer,
+    ClientOnly, // Server is considered offline for the whole run
+}
+
+#[derive(Debug)]
+pub struct TestbedEnv {
+    pub kind: TestbedKind,
+    // Config dir is used as the key identifying the testbed env
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/testbed/src/env.rs:17]:
```suggestion
    // Config dir is used as the key identifying the testbed env.
```

File: oxidation/libparsec/crates/testbed/src/env.rs (hunk 4 above)
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/testbed/src/env.rs:18]:
```suggestion
    /// Config dir is used as the key identifying the testbed env
    /// note testbed is designed to always run mocked in memory
```

Convert them to doc comment to provide that information to the dev

File: oxidation/libparsec/crates/testbed/src/types.rs [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    pub fn new(
+        id: &'static str,
+        root_signing_key: SigningKey,
+        devices: Vec<TestbedDeviceData>,
+        users: Vec<TestbedUserData>,
+        device_files: Vec<TestbedDeviceFileData>,
+    ) -> Self {
+        let mut hasher = crc32fast::Hasher::new();
+
+        // Remember that changing the order (or adding items) to the hasher change the result !
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/testbed/src/types.rs:42]: That seems normal behavior for an hash function (like `sha256` & co), I'm missing something ?
Maybe the reason of this comment is if this code block is change we may have to ensure the client & server run the same compiled code :thinking:

File: oxidation/libparsec/crates/testbed/src/types.rs [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+impl TestbedDeviceData {
+    pub(crate) fn new(
+        device_id: &DeviceID,
+        device_label: Option<DeviceLabel>,
+        signing_key: SigningKey,
+        local_symkey: SecretKey,
+        author: CertificateSignerOwned,
+        author_signkey: &SigningKey,
+        timestamp: DateTime,
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/testbed/src/types.rs:87]: Given the number of params, we may want to have a builder pattern for that in a latter time when we add some other params

File: oxidation/libparsec/crates/client_connection/Cargo.toml (hunk 2 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/client_connection/Cargo.toml:8]: You test the testbed, it's not redundant :thinking:

File: src/testbed.rs [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+            .collect()
+    }
+    #[getter]
+    fn crc(&self) -> u32 {
+        self.0.crc.to_owned()
+    }
+}
+
+#[pyfunction]
+pub(crate) fn test_new_testbed(
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply src/testbed.rs:155]: All the file is in `cfg(feature)` now :+1:

File: oxidation/libparsec/crates/testbed/src/coolorg.rs [hunk 9]
```diff
@@ -151,21 +158,21 @@ pub(crate) fn generate() -> TestbedTemplate {
     // Local devices
     let device_files = vec![
         TestbedDeviceFileData::new(
-            &alice1_id,
+            alice1_id,
             "P@ssw0rd.".to_owned(),
             SecretKey::from(hex!(
                 "323614fc6bd2d300f42d6731059f542f89534cdca11b2cb13d5a9a5a6b19b6ac"
             )),
         ),
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/testbed/src/coolorg.rs:166]: Could you create a macro for that ?

File: oxidation/libparsec/crates/tests_fixtures/src/testbed.rs [hunk 10]
```diff
@@ -51,27 +53,56 @@ macro_rules! run_in_testbed {
...
+    // - `AUTOSTART` or `1` or just defined to empty: auto start the serve
+    // - parsec://<domain>:<port>[?no_ssl=true],
+    //   <domain>:<port>,
+    //   http://<domain>:<port> or
+    //   https://<domain>:<port> consider the server is already started
+    let testbed_server = std::env::var("TESTBED_SERVER");
+
+    if let Err(e) = testbed_server {
+        // That means the env variable isn't set
+        println!("{e}, so the test is skipped !");
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/tests_fixtures/src/testbed.rs:95]:
```suggestion
        eprintln!("`TESTBED_SERVER` env var isn't set, skipping the test ! ({e})");
```

File: src/enumerate.rs [hunk 11]
```diff
@@ -14,8 +14,13 @@ use libparsec::{
 
 use crate::protocol::{ProtocolErrorFields, ProtocolResult};
 
+// #[non_exhaustive] macro must be set for every enum,
```
Comment (FirelightFlagboy) [inline src/enumerate.rs:17]:
```suggestion
// #[non_exhaustive] macro must be set for every enum like type,
```

Because here `non_exhaustive` is set on `rust` `struct` and this could be misleading as we're using them like an enum in the python context.

File: src/testbed.rs [hunk 12]
```diff
@@ -29,41 +31,38 @@ crate::binding_utils::gen_proto!(TestbedDeviceData, __deepcopy__);
...
     #[getter]
-    fn device_label(&self) -> Option<crate::ids::DeviceLabel> {
-        self.0
-            .device_label
-            .as_ref()
-            .map(|x| crate::ids::DeviceLabel(x.clone()))
+    fn device_label(&self) -> Option<DeviceLabel> {
+        self.0.device_label.as_ref().map(|x| DeviceLabel(x.clone()))
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:41]: spacing

File: src/testbed.rs [hunk 13]
```diff
@@ -29,41 +31,38 @@ crate::binding_utils::gen_proto!(TestbedDeviceData, __deepcopy__);
 #[pymethods]
 impl TestbedDeviceData {
     #[getter]
-    fn device_id(&self) -> crate::ids::DeviceID {
-        crate::ids::DeviceID(self.0.device_id.clone())
+    fn device_id(&self) -> DeviceID {
+        DeviceID(self.0.device_id.clone())
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:37]: spacing

File: src/testbed.rs [hunk 14]
```diff
@@ -29,41 +31,38 @@ crate::binding_utils::gen_proto!(TestbedDeviceData, __deepcopy__);
...
+    fn device_label(&self) -> Option<DeviceLabel> {
+        self.0.device_label.as_ref().map(|x| DeviceLabel(x.clone()))
     }
     #[getter]
-    fn signing_key(&self) -> crate::api_crypto::SigningKey {
-        crate::api_crypto::SigningKey(self.0.signing_key.clone())
+    fn signing_key(&self) -> SigningKey {
+        SigningKey(self.0.signing_key.clone())
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:45]: spacing

File: src/testbed.rs [hunk 15]
```diff
@@ -29,41 +31,38 @@ crate::binding_utils::gen_proto!(TestbedDeviceData, __deepcopy__);
...
-    fn local_symkey(&self) -> crate::api_crypto::SecretKey {
-        crate::api_crypto::SecretKey(self.0.local_symkey.clone())
+    fn local_symkey(&self) -> SecretKey {
+        SecretKey(self.0.local_symkey.clone())
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:49]: spacing

File: src/testbed.rs [hunk 16]
```diff
@@ -29,41 +31,38 @@ crate::binding_utils::gen_proto!(TestbedDeviceData, __deepcopy__);
...
-        crate::data::DeviceCertificate(self.0.certif.clone())
+    fn certif(&self) -> DeviceCertificate {
+        DeviceCertificate(self.0.certif.clone())
     }
     #[getter]
-    fn raw_certif<'p>(&self, py: Python<'p>) -> &'p PyBytes {
+    fn raw_certif<'py>(&self, py: Python<'py>) -> &'py PyBytes {
         PyBytes::new(py, &self.0.raw_certif)
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:57]: spacing

File: src/testbed.rs [hunk 17]
```diff
@@ -29,41 +31,38 @@ crate::binding_utils::gen_proto!(TestbedDeviceData, __deepcopy__);
...
+    fn local_symkey(&self) -> SecretKey {
+        SecretKey(self.0.local_symkey.clone())
     }
     #[getter]
-    fn certif(&self) -> crate::data::DeviceCertificate {
-        crate::data::DeviceCertificate(self.0.certif.clone())
+    fn certif(&self) -> DeviceCertificate {
+        DeviceCertificate(self.0.certif.clone())
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:53]: spacing

File: src/testbed.rs [hunk 18]
```diff
@@ -72,51 +71,49 @@ crate::binding_utils::gen_proto!(TestbedUserData, __deepcopy__);
 #[pymethods]
 impl TestbedUserData {
     #[getter]
-    fn user_id(&self) -> crate::ids::UserID {
-        crate::ids::UserID(self.0.user_id.clone())
+    fn user_id(&self) -> UserID {
+        UserID(self.0.user_id.clone())
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:77]: spacing

File: src/testbed.rs [hunk 19]
```diff
@@ -72,51 +71,49 @@ crate::binding_utils::gen_proto!(TestbedUserData, __deepcopy__);
...
-    fn human_handle(&self) -> Option<crate::ids::HumanHandle> {
+    fn human_handle(&self) -> Option<HumanHandle> {
         self.0
             .human_handle
             .as_ref()
             .map(|x| crate::ids::HumanHandle(x.clone()))
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:84]: spacing

File: src/testbed.rs [hunk 20]
```diff
@@ -72,51 +71,49 @@ crate::binding_utils::gen_proto!(TestbedUserData, __deepcopy__);
...
-    fn private_key(&self) -> crate::api_crypto::PrivateKey {
-        crate::api_crypto::PrivateKey(self.0.private_key.clone())
+    fn private_key(&self) -> PrivateKey {
+        PrivateKey(self.0.private_key.clone())
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:88]: spacing

File: src/testbed.rs [hunk 21]
```diff
@@ -72,51 +71,49 @@ crate::binding_utils::gen_proto!(TestbedUserData, __deepcopy__);
...
-    fn profile(&self) -> crate::enumerate::UserProfile {
-        crate::enumerate::UserProfile(self.0.profile.clone())
+    fn profile(&self) -> &'static PyObject {
+        UserProfile::from_profile(self.0.profile)
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:92]: spacing

File: src/testbed.rs [hunk 22]
```diff
@@ -72,51 +71,49 @@ crate::binding_utils::gen_proto!(TestbedUserData, __deepcopy__);
...
-    fn user_manifest_id(&self) -> crate::ids::EntryID {
-        crate::ids::EntryID(self.0.user_manifest_id.clone())
+    fn user_manifest_id(&self) -> EntryID {
+        EntryID(self.0.user_manifest_id.clone())
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:96]: spacing

File: src/testbed.rs [hunk 23]
```diff
@@ -72,51 +71,49 @@ crate::binding_utils::gen_proto!(TestbedUserData, __deepcopy__);
...
-    fn user_manifest_key(&self) -> crate::api_crypto::SecretKey {
-        crate::api_crypto::SecretKey(self.0.user_manifest_key.clone())
+    fn user_manifest_key(&self) -> SecretKey {
+        SecretKey(self.0.user_manifest_key.clone())
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:100]: spacing

File: src/testbed.rs [hunk 24]
```diff
@@ -72,51 +71,49 @@ crate::binding_utils::gen_proto!(TestbedUserData, __deepcopy__);
...
-    fn certif(&self) -> crate::data::UserCertificate {
-        crate::data::UserCertificate(self.0.certif.clone())
+    fn certif(&self) -> UserCertificate {
+        UserCertificate(self.0.certif.clone())
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:104]: spacing

File: src/testbed.rs [hunk 25]
```diff
@@ -72,51 +71,49 @@ crate::binding_utils::gen_proto!(TestbedUserData, __deepcopy__);
...
-    fn raw_certif<'p>(&self, py: Python<'p>) -> &'p PyBytes {
+    fn raw_certif<'py>(&self, py: Python<'py>) -> &'py PyBytes {
         PyBytes::new(py, &self.0.raw_certif)
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:108]: spacing

File: src/testbed.rs [hunk 26]
```diff
@@ -125,12 +122,12 @@ crate::binding_utils::gen_proto!(TestbedTemplate, __deepcopy__);
 #[pymethods]
 impl TestbedTemplate {
     #[getter]
-    fn id(&self) -> String {
-        self.0.id.to_owned()
+    fn id(&self) -> &str {
+        self.0.id
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:128]: spacing

File: src/testbed.rs [hunk 27]
```diff
@@ -125,12 +122,12 @@ crate::binding_utils::gen_proto!(TestbedTemplate, __deepcopy__);
...
-    fn root_signing_key(&self) -> crate::api_crypto::SigningKey {
-        crate::api_crypto::SigningKey(self.0.root_signing_key.clone())
+    fn root_signing_key(&self) -> SigningKey {
+        SigningKey(self.0.root_signing_key.clone())
     }
     #[getter]
```
Comment (FirelightFlagboy) [inline src/testbed.rs:132]: spacing

File: src/testbed.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply src/testbed.rs:155]: :+1:

File: .pre-commit-config.yaml [hunk 28]
```diff
@@ -146,9 +146,7 @@ repos:
             --exclude=libparsec_bindings_electron,
             --,
             --deny=warnings,
-            --deny=clippy::self_named_module_files,
```
Comment (FirelightFlagboy) [inline .pre-commit-config.yaml:149]: Nop, we keep that

**Review by FirelightFlagboy — CHANGES_REQUESTED** [review]

File: oxidation/libparsec/crates/tests_fixtures/src/testbed.rs (hunk 10 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/tests_fixtures/src/testbed.rs:95]: Why `stderr` ? This is not an error

File: oxidation/libparsec/crates/testbed/src/coolorg.rs (hunk 9 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/testbed/src/coolorg.rs:166]: Why ?

File: .pre-commit-config.yaml (hunk 28 above)
Comment (TimeEngineer) [reply .pre-commit-config.yaml:149]: This lint is broken, @touilleMan can't have his module named `types` because `libparsec_types` already uses this name

Comment (FirelightFlagboy) [reply .pre-commit-config.yaml:149]: So it's a bug, do we have opened an issue on clippy ?

File: oxidation/libparsec/crates/tests_fixtures/src/testbed.rs (hunk 10 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/tests_fixtures/src/testbed.rs:95]: the value of `e` can change, It by doing the review I don't known it's value beforehand

File: oxidation/libparsec/crates/testbed/src/coolorg.rs (hunk 9 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/testbed/src/coolorg.rs]: To factorize code, to have less text to read :shrug:

Comment (TimeEngineer) [reply oxidation/libparsec/crates/testbed/src/coolorg.rs]: 8 * 3 = 24 lines into 4 * 3 + 10 = 22 lines + a macro only for that ?

File: .pre-commit-config.yaml (hunk 28 above)
Comment (TimeEngineer) [reply .pre-commit-config.yaml:149]: Reported: https://github.com/rust-lang/rust-clippy/issues/10271

Comment (FirelightFlagboy) [reply .pre-commit-config.yaml:149]: :+1:

File: oxidation/libparsec/crates/testbed/src/coolorg.rs (hunk 9 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/testbed/src/coolorg.rs]: So no macro

File: oxidation/libparsec/crates/testbed/Cargo.toml [hunk 29]
```diff
@@ -0,0 +1,16 @@
...
+libparsec_crypto = { path = "../crypto" }
+libparsec_types = { path = "../types" }
+libparsec_client_types = { path = "../client_types" }
+
+hex-literal = "0.3.3"
+crc32fast = "1.3.2"
+serde = { version = "1.0.147", features = ["derive"] }
+rmp-serde = "1.1.1"
+lazy_static = "1.4.0"
+reqwest = { version = "0.11.13", features = ["native-tls-vendored"] }
```
Comment (touilleMan) [inline oxidation/libparsec/crates/testbed/Cargo.toml:16]: Why the explicit feature is needed ?

Feature unification with crate `libparsec_client_connection` makes reqwest have this feature enabled no matter what

Comment (TimeEngineer) [reply oxidation/libparsec/crates/testbed/Cargo.toml:16]: It's more explicit

File: oxidation/libparsec/crates/platform_device_loader/src/testbed.rs (hunk 3 above)
Comment (touilleMan) [reply oxidation/libparsec/crates/platform_device_loader/src/testbed.rs]: The function taking a callback is needed because we need to keep the lock on `env.persistence_store` for the whole time the callback is called.

File: oxidation/libparsec/crates/testbed/src/types.rs (hunk 6 above)
Comment (touilleMan) [reply oxidation/libparsec/crates/testbed/src/types.rs:46]: > Maybe the reason of this comment is if this code block is change we may have to ensure the client & server run the same compiled code 🤔

Yes, basically changing the order is a breaking change that is not visible as long as the same codebase is used for the testbed server and the tests. As you say changing order for hash is pretty common, but we want to emphases anyway given how simple it is to add a component or reorder something here.

File: oxidation/libparsec/crates/testbed/src/types.rs [hunk 30]
```diff
@@ -0,0 +1,206 @@
...
+    pub user_manifest_id: EntryID,
+    pub user_manifest_key: SecretKey,
+    pub certif: UserCertificate,
+    pub raw_certif: Vec<u8>,
+    pub raw_redacted_certif: Vec<u8>,
+    pub crc: u32,
+}
+
+impl TestbedUserData {
+    #[allow(clippy::too_many_arguments)]
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply oxidation/libparsec/crates/testbed/src/types.rs:139]: Those parameters are all required to create those objects. I don't think a builder is good fit here given we always want to be explicit about the thing we put in those types and we always create them in the same scenario (i.e. in env template initialization).

File: oxidation/libparsec/crates/tests_fixtures/src/testbed.rs [hunk 31]
```diff
@@ -0,0 +1,168 @@
...
+/// across all tests.
+/// Alternatively, `TESTBED_SERVER` environ variable can be passed to use an
+/// external server.
+fn ensure_testbed_server_is_started() -> BackendAddr {
+    if let Ok(raw) = std::env::var("TESTBED_SERVER") {
+        let addr = BackendAddr::from_any(&raw)
+            .expect("Invalid value in `TESTBED_SERVER` environ variable");
+        println!("Using already started testbed server at {}", &addr);
+        return addr;
+    }
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply oxidation/libparsec/crates/tests_fixtures/src/testbed.rs]: That's not needed: the test will fail as soon as the server is needed when doing the first `/testbed/new` query

File: oxidation/libparsec/crates/testbed/Cargo.toml (hunk 29 above)
Comment (touilleMan) [reply oxidation/libparsec/crates/testbed/Cargo.toml:16]: We don't care if tls is vendored here (given this is crate that is never shipped in release), we just want reqwest.
On the other hand we don't want this feature to clash with the ones set in `libparsec_platform_client_connection` where the portability question is actually important.

For instance: does `native-tls-vendored` feature available when building for web ? do we want/need it on windows where TLS support is not provided by openssl ?
Those are question we don't need to answer here but in `libparsec_platform_client_connection`

File: oxidation/libparsec/src/lib.rs [hunk 32]
```diff
@@ -9,11 +9,12 @@ pub use libparsec_protocol as protocol;
 #[cfg(not(target_arch = "wasm32"))]
 pub use libparsec_types as types;
 
+#[cfg(feature = "test-utils")]
+pub use libparsec_testbed as testbed;
```
Comment (touilleMan) [inline oxidation/libparsec/src/lib.rs:13]: why is this needed ?

we want the stuff exposed by testbed to be part of the `libparsec` public api

Comment (FirelightFlagboy) [reply oxidation/libparsec/src/lib.rs:13]: We want the testbed api to be present in a production build ?

File: oxidation/bindings/generator/templates/binding_electron_meths.rs.j2 [hunk 33]
```diff
@@ -466,6 +478,12 @@ fn {{ meth.name }}(mut cx: FunctionContext) -> JsResult<JsPromise> {
 
 #[allow(unused_imports)]
 use neon::{prelude::*, types::buffer::TypedArray};
+#[allow(dead_code)]
+const I32_MAX: f64 = i32::MAX as f64;
+#[allow(dead_code)]
+const I32_MIN: f64 = i32::MIN as f64;
+const U32_MAX: f64 = u32::MAX as f64;
```
Comment (touilleMan) [inline oxidation/bindings/generator/templates/binding_electron_meths.rs.j2:485]: no need for `#[allow(dead_code)]` for U32_* ?

Don't we want to just inline `i32::MIN as f64` in the callsite to not have to wonder about that ?

File: oxidation/libparsec/src/lib.rs (hunk 32 above)
Comment (TimeEngineer) [reply oxidation/libparsec/src/lib.rs:13]: If you can find a way that the `front generator` can use testbed only in test :man_shrugging:

File: oxidation/bindings/generator/templates/binding_electron_meths.rs.j2 (hunk 33 above)
Comment (TimeEngineer) [reply oxidation/bindings/generator/templates/binding_electron_meths.rs.j2:485]: `U32_* is already used`

File: oxidation/libparsec/crates/client_high_level_api/Cargo.toml [hunk 34]
```diff
@@ -14,6 +14,7 @@ libparsec_crypto = { path = "../crypto" }
 libparsec_platform_async = { path = "../platform_async" }
 libparsec_platform_device_loader = { path = "../platform_device_loader" }
 libparsec_protocol = { path = "../protocol" }
+libparsec_testbed = { path = "../testbed" }
```
Comment (touilleMan) [inline oxidation/libparsec/crates/client_high_level_api/Cargo.toml:17]: `libparsec_testbed` should not be present in release, so we shouldn't have this unconditional add here !

Comment (touilleMan) [reply oxidation/libparsec/crates/client_high_level_api/Cargo.toml:17]: (see what is done for e.g. `libparsec_platform_device_loader`)

File: oxidation/libparsec/src/lib.rs (hunk 32 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/src/lib.rs:13]: > We have similar issue in the Python binding

Except for the typing, we can include or not elements at the compilation without providing a mocked function 
see https://github.com/Scille/parsec-cloud/blob/31451b41960b03270a8d025998eedf8c6f05e327/src/lib.rs#L31-L32

File: oxidation/libparsec/crates/testbed/src/types.rs (hunk 30 above)
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/testbed/src/types.rs:139]: I don't understand how `we always want to be explicit about the thing we put in those types` contradicts with the builder pattern method :thinking:

File: oxidation/libparsec/crates/testbed/Cargo.toml (hunk 29 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/testbed/Cargo.toml]: Done

File: oxidation/libparsec/crates/testbench/src/coolorg.rs (hunk 1 above)
Comment (touilleMan) [reply oxidation/libparsec/crates/testbench/src/coolorg.rs]: better do that in another PR (typically when we will add further stuff to the template like manifests) to have this one merged asap

File: oxidation/libparsec/crates/platform_device_loader/Cargo.toml [hunk 35]
```diff
@@ -3,17 +3,25 @@ name = "libparsec_platform_device_loader"
 version = "0.0.0"
 edition = "2021"
 
+[features]
+test-testbed-support = [
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply oxidation/libparsec/crates/platform_device_loader/Cargo.toml:7]: god for it 👍

**Review by FirelightFlagboy — APPROVED** [review]
